keps/sig-etcd/4743-kuberernetes-etcd-interface/README.md

```diff
@@ -0,0 +1,349 @@
+- [KEP-4743 The Kubernetes-etcd interface](#kep-4743-the-kubernetes-etcd-interface)
+  - [Summary](#summary)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Kubernetes Backporting a etcd client patch version](#kubernetes-backporting-a-etcd-client-patch-version)
+    - [Making Changes/Patching Bugs in the Interface](#making-changespatching-bugs-in-the-interface)
+    - [Kubernetes Leveraging New etcd Functionality](#kubernetes-leveraging-new-etcd-functionality)
+  - [Code location](#code-location)
+  - [The Interface](#the-interface)
+    - [KV interface](#kv-interface)
+    - [Design considerations](#design-considerations)
+    - [Watch interface](#watch-interface)
+    - [Design considerations](#design-considerations-1)
+  - [Alternatives](#alternatives)
+    - [Code location](#code-location-1)
+      - [Part of the etcd Client Struct](#part-of-the-etcd-client-struct)
+      - [New Package in etcd Repository](#new-package-in-etcd-repository)
+      - [New Repository under etcd-io](#new-repository-under-etcd-io)
+<!-- /toc -->
+# KEP-4743 The Kubernetes-etcd interface
+
+## Summary 
+
+This design proposal introduces an etcd-Kubernetes interface to be added to the 
+etcd client and adopted by Kubernetes. This interface aims to create a clear and
+standardized contract between the two projects, codifying the interactions
+outlined in the [Implicit Kubernetes-ETCD Contract]. By formalizing this contract,
+we will improve the testability of both Kubernetes and etcd, prevent common
+errors in their interaction, and establish a framework for the future evolution 
+of this critical contract.
+
+[Implicit Kubernetes-ETCD Contract]: https://docs.google.com/document/d/1NUZDiJeiIH5vo_FMaTWf0JtrQKCx0kpEaIIuPoj9P6A/edit#heading=h.tlkin1a8b8bl
+
+### Goals
+
+* **Improved Testability:** Enable thorough testing of etcd and Kubernetes 
+  interactions through a well-defined interface, as envisioned in [#15820].
+* **Error Prevention:** Reduce incorrect contract usage, addressing issues like Kubernetes [#110210].
+* **Reviewable Changes:** Make contract modifications easily reviewable and trackable, ensuring a transparent and collaborative evolution.
+
+[#15820]: https://github.com/etcd-io/etcd/issues/15820
+[#110210]: https://github.com/kubernetes/kubernetes/issues/110210
+
+In scope
+* [etcd3 store]: The primary Kubernetes object storage interface.
+* [Master leases]: Lease management for Kubernetes control plane components (utilizing the [etcd3 store])
+
+[etcd3 store]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
+[Master leases]: https://github.com/kubernetes/kubernetes/blob/dae1859c896d742de1ee60a349475f8e28b61995/pkg/controlplane/reconcilers/lease.go#L47-L66
+
+### Non-Goals
+* **Alternative Interface Implementations:** This KEP focuses on defining the interface, not on providing alternative implementations for etcd.
```

Comment:
I agree with this.  Can we also include in this document a statement about how we should requests for features in the interface that non-etcd backends can support but etcd cannot?

Reply:
Sound great, let me link the official api-machinery policy on this.